modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/mqtt2opcua/JsonToOpcUAConverter.java

```diff
@@ -363,28 +251,247 @@ public JsonToOpcUAConverter(final @NotNull OpcUaClient client) throws UaExceptio
                 // DiagnosticInfo is too complex for now
                 throw new NotImplementedException();
             case ExtensionObject:
-                if (binaryEncodingId == null) {
-                    throw new RuntimeException("Binary encoding id was null for nested struct. ");
-                }
-
-                final Map<String, FieldType> fields = getStructureInformation(binaryEncodingId);
-                final Struct.Builder builder = Struct.builder("CustomStruct"); // apparently the name is not important
-
-                for (final Map.Entry<String, FieldType> entry : fields.entrySet()) {
-                    final String key = entry.getKey();
-                    final FieldType fieldType = entry.getValue();
-                    final JsonNode nestedObjectNode = jsonNode.get(key);
-                    if (nestedObjectNode == null) {
-                        throw new RuntimeException("No nested json was found for key '" + key + "'.");
-                    }
-                    final Object parsed = parseToOpcUACompatibleObject(nestedObjectNode, fieldType);
-                    builder.addMember(key, parsed);
-                }
-                return builder.build();
+                return extractExtensionObject(jsonNode, binaryEncodingId);
         }
         throw createException(jsonNode, builtinDataType.name());
     }
 
+    private @NotNull Struct extractExtensionObject(
+            @NotNull final JsonNode jsonNode, @Nullable final NodeId binaryEncodingId) {
+        if (binaryEncodingId == null) {
+            throw new RuntimeException("Binary encoding id was null for nested struct. ");
```

Comment:
Isn't that rather an IllegalArgumentException?
Also, why @Nullable and then throw an exception when it's null?

I'd still rather go with using optional as return type here and not throw an exception, it would make the flow more explicit.

Reply:
This is an exceptional case as it will hinder the parsing and in the end writing of the data. Optional.absent would not be a legit case. But yeah Runtime is too generic.